[AArch64] Implement INIT/ADJUST_TRAMPOLINE

ceseo opened the pull request:
Add support for llvm.init.trampoline and llvm.adjust.trampoline intrinsics for
AArch64.

Fixes https://github.com/llvm/llvm-project/issues/65573
Fixes https://github.com/llvm/llvm-project/issues/76927
Fixes https://github.com/llvm/llvm-project/issues/83555
Updates https://github.com/llvm/llvm-project/pull/66157

kiranchandramohan commented:
@ceseo Thanks for this patch.
Did you see the related patch in Phabricator? https://reviews.llvm.org/D149848

ceseo commented:
> @ceseo Thanks for this patch. Did you see the related patch in Phabricator? https://reviews.llvm.org/D149848

I missed that one. Thanks for pointing that out.

So, I think I agree that it is better to do the same as gcc does. The code will probably be very untidy, since we have to write the actual instructions (like in the x86 implementation), but it should work too.

The only observation I have is that, whatever the approach is, we cannot use `x18` for the nest parameter outside Linux because both macOS and Windows ABIs already assign this register for something else.

If the submitter of that patch wants to do this, I'm OK with that since I'll be away for a month or so.

compnerd commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5946,6 +5946,56 @@ static SDValue LowerFLDEXP(SDValue Op, SelectionDAG &DAG) {
+
+  // Note: x18 cannot be used for the Nest parameter on Windows and macOS.
+  if (Subtarget->isTargetDarwin() || Subtarget->isTargetWindows())
+    report_fatal_error("INIT_TRAMPOLINE operation is only supported on Linux.");
```
Windows ARM64 supports x16, x17 as intra-procedural scratch. AIUI, GCC does not have `__trampoline_setup` for ARM64, and so, I would say that we should try to adjust the ABI here to make this possible to use on Windows as well.  Is this currently defined in an accepted specification or is this an extension being added?

ceseo replied:
Per the Arm PCS:

"The role of register r18 is platform specific. If a platform ABI has need of a dedicated general purpose register to carry inter-procedural state (for example, the thread context) then it should use this register for that purpose. If the platform ABI has no such requirements, then it should use r18 as an additional temporary register. The platform ABI specification must document the usage for this register."

Both macOS and Windows use this as the Platform Register and LLVM documents this in `AArch64CallingConvention.td`.

ceseo replied:
Also, something just came to my mind. This kind of trampoline for nested function calls requires executable stack. Doesn't DEP block that on Windows?

ceseo commented on compiler-rt/lib/builtins/README.txt:
```diff
@@ -272,6 +272,11 @@ switch32
+// This function generates a custom trampoline function with the specific
+// realFunc and localsPtr values.
+void __trampoline_setup(uint32_t* trampOnStack, int trampSizeAllocated,
+                                const void* realFunc, void* localsPtr);
```
OK, will do. Strange that `git clang-format` didn't catch this for me.

giordano commented:
Probably fixes #56625 too?

ceseo commented:
> Probably fixes #56625 too? _**Edit**_: not yet, if this is limited to Linux

Yes, we need to sort out the nest parameter register for Windows and macOS. Both use `x18` as the platform register. So this is Linux only for now.

ceseo commented:
Any other comments on this?

ceseo commented:
We will try another approach that avoids executable stack when procedure pointers support is fully implemented.

vzakhari commented:
I think this change is useful on its own for aarch64 Linux, and I think it can be merged as is.

kiranchandramohan commented:
@ceseo What is your plan here? Do you have a different patch?

ceseo commented:
Not for now. I'm waiting until the support for procedure pointers is fully implemented and then I'll give it another shot to avoid executable stack.

If you think this is useful as is until then, please feel free to merge it.

kiranchandramohan commented:
@ceseo Could you reopen the PR?

MaskRay commented:
Description:

> Updates `https://github.com/llvm/llvm-project/issues/27174`

27174 is an AMDGPU issue. Is there a typo?

ceseo commented:
> Description:
> 
> > Updates `https://github.com/llvm/llvm-project/issues/27174`
> 
> 27174 is an AMDGPU issue. Is there a typo?

Sorry, I had the number from an issue in the Julia repo in my mind. I'll fix that. This PR actually fixes other issues as well. I'll reference those.

ceseo committed:
[AArch64] Implement INIT/ADJUST_TRAMPOLINE

Add support for llvm.init.trampoline and llvm.adjust.trampoline intrinsics for
AArch64.

Fixes #65573
Fixes #76927
Fixes #83555
Updates #66157

vzakhari commented on compiler-rt/lib/builtins/trampoline_setup.c:
```diff
@@ -41,3 +41,30 @@ COMPILER_RT_ABI void __trampoline_setup(uint32_t *trampOnStack,
+// which loads x18 with a pointer to the outer function's locals
+// and then jumps to the target nested function.
+
+#if __aarch64__
```
I would suggest adding `defined(__linux__)` guard as well.

MaskRay replied:
This can be `defined(__aarch64__) && defined(__ELF__)` since other ELF OSes likely need this as well.

ceseo replied:
There is already an OS guard in the lowering code. Does it need to be done here as well? I don't think anything else call this code.

vzakhari replied:
I prefer the library implementation being guarded as well, so that its limitations are clear in the library source code.

vzakhari commented on compiler-rt/lib/builtins/trampoline_setup.c:
```diff
@@ -41,3 +41,30 @@ COMPILER_RT_ABI void __trampoline_setup(uint32_t *trampOnStack,
+  trampOnStack[4] = (uint32_t)localsPtr;
+
+  // clear instruction cache
+  __clear_cache(trampOnStack, &trampOnStack[10]);
```
```suggestion
  __clear_cache(trampOnStack, &trampOnStack[5]);
```

ceseo commented:
The address loading is wrong as well. I'll fix that and resubmit.

vzakhari commented on compiler-rt/lib/builtins/trampoline_setup.c:
```diff
@@ -41,3 +41,30 @@ COMPILER_RT_ABI void __trampoline_setup(uint32_t *trampOnStack,
+  trampOnStack[0] = 0x580000b1; // ldr x17, realFunc
+  trampOnStack[1] = 0x580000d2; // ldr x18, localsPtr
+  trampOnStack[2] = 0xd61f0220; // br x17
+  trampOnStack[3] = (uint32_t)realFunc;
```
The casts seem to be truncating.

vzakhari commented:
It is also worth adding a LIT test for the ISEL.

DavidTruby reviewed (commented):
I had a go at trying this out for a case where we generate the trampoline in flang but we currently don't have the -rtlib=compiler-rt flag and it looks like this will require compiler-rt over libgcc_s? I'll try to quickly implement -rtlib so I can see if this works for our flang cases.

ceseo commented:
> I had a go at trying this out for a case where we generate the trampoline in flang but we currently don't have the -rtlib=compiler-rt flag and it looks like this will require compiler-rt over libgcc_s? I'll try to quickly implement -rtlib so I can see if this works for our flang cases.

I noticed this today as well. It wasn't happening when I first submitted the patch a long time ago. Did something change in compiler-rt that may affect this?

DavidTruby commented:
> I noticed this today as well. It wasn't happening when I first submitted the patch a long time ago. Did something change in compiler-rt that may affect this?

I don't think anything changed; we don't link to compiler-rt on Linux, instead we link to libgcc_s. Clang does this by default too but has a flag to choose compiler-rt instead. I don't have the __start_trampoline symbol in libgcc_s on my system so I guess we'll need to require compiler-rt for this.

I don't think that invalidates the patch, at least as a stopgap until we can get non-executable stack trampolines working. On x86 LLVM doesn't require an external function for this, it just does the code inline. I'm guessing that's quite difficult to do here?

DavidTruby commented:
I've added the -rtlib flag in #99058. With that change and this one, I am able to build quantum-espresso with `-rtlib=compiler-rt` and have it pass tests, which previously failed due to the missing trampoline implementation.

ceseo commented on compiler-rt/lib/builtins/README.txt:
```diff
@@ -272,6 +272,11 @@ switch32
 
+// This function generates a custom trampoline function with the specific
+// realFunc and localsPtr values.
+void __trampoline_setup(uint32_t* trampOnStack, int trampSizeAllocated,
+                        const void* realFunc, void* localsPtr);
```
The latter.

ceseo commented:
> > I noticed this today as well. It wasn't happening when I first submitted the patch a long time ago. Did something change in compiler-rt that may affect this?
> 
> I don't think anything changed; we don't link to compiler-rt on Linux, instead we link to libgcc_s. Clang does this by default too but has a flag to choose compiler-rt instead. I don't have the __start_trampoline symbol in libgcc_s on my system so I guess we'll need to require compiler-rt for this.
> 
> I don't think that invalidates the patch, at least as a stopgap until we can get non-executable stack trampolines working. On x86 LLVM doesn't require an external function for this, it just does the code inline. I'm guessing that's quite difficult to do here?

Correct.

vzakhari reviewed (approved):
Thank you! LGTM

MaskRay commented on llvm/test/CodeGen/AArch64/trampoline.ll:
```diff
@@ -0,0 +1,31 @@
+; UNSUPPORTED: darwin, system-windows
+
+declare void @llvm.init.trampoline(i8*, i8*, i8*);
+declare i8* @llvm.adjust.trampoline(i8*);
```
Avoid typed pointers like `i8*`. Use opaque pointers `ptr`

MaskRay commented on llvm/test/CodeGen/AArch64/trampoline.ll:
```diff
@@ -0,0 +1,31 @@
+declare void @llvm.init.trampoline(i8*, i8*, i8*);
+declare i8* @llvm.adjust.trampoline(i8*);
+
+define i64 @func(i64* nest %ptr, i64 %val)
```
The prevailing style is to place `{` at the end of the previous line

MaskRay commented:
> Updates #27174

Can you change the description? It will be used as the default commit message when you click "Squash and merge".
git commit messages are completely ignored.

ceseo commented:
> > Updates #27174
> 
> Can you change the description? It will be used as the default commit message when you click "Squash and merge". git commit messages are completely ignored.

OK. I'll fix the test and the description and resubmit.

ceseo commented:
I also simplified the test.

DavidTruby reviewed (approved):
LGTM, but wait for @MaskRay to approve. 

Could you update the NOTE about aarch64 in https://github.com/llvm/llvm-project/blob/main/flang/docs/InternalProcedureTrampolines.md to note that this is implemented now, but requires compiler-rt (and -rtlib=compiler-rt)? That can be in a separate NFC patch without review after -rtlib=compiler-rt goes in, if you merge this first.

ceseo commented:
Ping?

ceseo commented:
@kiranchandramohan do you want this in for LLVM 19, or should we leave it for LLVM 20?

kiranchandramohan commented:
> @kiranchandramohan do you want this in for LLVM 19, or should we leave it for LLVM 20?

Yes, it will be great to have this for LLVM 19 since this enables an important HPC application.

MaskRay commented on llvm/test/CodeGen/AArch64/trampoline.ll:
```diff
@@ -0,0 +1,20 @@
+; RUN: llc -mtriple=aarch64-- < %s | FileCheck %s
```
`-mtriple=aarch64` specifies a generic ELF triple (yes, ELF is the default:) )

`; UNSUPPORTED: darwin, system-windows` should not be needed.

ceseo replied:
OK, I'll remove that.

MaskRay commented on compiler-rt/lib/builtins/trampoline_setup.c:
```diff
@@ -41,3 +41,45 @@ COMPILER_RT_ABI void __trampoline_setup(uint32_t *trampOnStack,
+// and then jumps to the target nested function.
+// Note: x18 is a reserved platform register on Windows and macOS.
+
+#if defined(__aarch64__) && !defined(__APPLE__) && !defined(_WIN64)
```
Perhaps `defined(__ELF__)` is slightly better and clearly excludes Apple (Mach-O) and Windows (PE/COFF).

ceseo replied:
OK, makes sense.

MaskRay reviewed (approved):
LGTM.

Ideally @compnerd could remove the "Request changes".

I agree that this will be nice to be merged into relase/19.x

ceseo commented:
> Yes, it will be great to have this for LLVM 19 since this enables an important HPC application.

I'll just make the small changes requested by @MaskRay and resubmit the final version.

ceseo commented:
@kiranchandramohan I'll commit this once all the checks pass.

ceseo commented:
I will merge this PR. The Windows failure is due to something else and it was fixed by 7b619730d9e2bcc2ab1588574ed94ae19c4b8654

eschnett commented:
It looks like LLVM 19 will support this.

vchuravy commented:
As a note GCC v14 has implemented heap based trampolines (using x18 on Linux and x16 on Apple).

https://gcc.gnu.org/onlinedocs/gcc-14.1.0/gcc/Code-Gen-Options.html#index-ftrampoline-impl

vtjnash commented:
heap-based trampolines is a higher-level (clang or flang) concern, not llvm, since llvm's intrinsics just takes a pointer to some memory, but doesn't care where/how it got allocated
